crates/bevy_animation/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+                        let mut keyframe_timestamps = curve.keyframe_timestamps.iter().enumerate();
+                        let mut step_start = keyframe_timestamps.next().unwrap();
+                        if elapsed < *step_start.1 {
+                            continue;
+                        }
+                        for next in keyframe_timestamps {
+                            if *next.1 > elapsed {
+                                break;
+                            }
+                            step_start = next;
+                        }
+                        if step_start.0 == curve.keyframe_timestamps.len() - 1 {
+                            // This curve is finished
+                            continue;
+                        }
+                        let step_end = curve.keyframe_timestamps[step_start.0 + 1];
+                        let lerp = (elapsed - *step_start.1) / (step_end - step_start.1);
+
+                        // Apply the keyframe
+                        match &curve.keyframes {
+                            Keyframes::Rotation(keyframes) => {
+                                let rot_start = keyframes[step_start.0];
+                                let mut rot_end = keyframes[step_start.0 + 1];
+                                // Choose the smallest angle for the rotation
+                                if rot_end.dot(rot_start) < 0.0 {
+                                    rot_end = -rot_end;
+                                }
+                                // Rotations are using a spherical linear interpolation
+                                transform.rotation = Quat::from_array(rot_start.normalize().into())
+                                    .slerp(Quat::from_array(rot_end.normalize().into()), lerp);
+                            }
+                            Keyframes::Translation(keyframes) => {
+                                let translation_start = keyframes[step_start.0];
+                                let translation_end = keyframes[step_start.0 + 1];
+                                let result = translation_start.lerp(translation_end, lerp);
+                                transform.translation = result;
+                            }
+                            Keyframes::Scale(keyframes) => {
+                                let scale_start = keyframes[step_start.0];
+                                let scale_end = keyframes[step_start.0 + 1];
+                                let result = scale_start.lerp(scale_end, lerp);
+                                transform.scale = result;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Adds animation support to an app
+#[derive(Default)]
+pub struct AnimationPlugin {}
+
+impl Plugin for AnimationPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_asset::<AnimationClip>()
+            .add_system(animation_player);
```

Comment:
I try to put everything in the `Update` stage by default... it could make sense to use the `PostUpdate` here as it may need to happen after an update from the user, but it also updates a lot of transforms which I prefer happens in the `Update` stage...

Reply:
I also think this should be in `PostUpdate` right now (with`.before(TransformSystem::TransformPropagate)` and `.after(HierarchySystem::ParentUpdate)` labels). `Update` (in a non-stageless world) is the place where users should put their game logic and have it "just work". Users shouldn't need to think about resolving order ambiguity between their "animation logic" and an internal "animation system". With the current approach, the ambiguity could result in skipping processing user-provided "animation input" until the next frame.

Reply:
done 👍

Reply:
This will definitely get interesting with stageless. I think we'll need some sort of implicit "update" label that gets added by default.